wave_utils/include/wave/utils/measurement.hpp

```diff
@@ -0,0 +1,312 @@
+#ifndef WAVE_UTILS_MEASUREMENT_HPP
+#define WAVE_UTILS_MEASUREMENT_HPP
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/composite_key.hpp>
+#include <boost/version.hpp>
+#include <iterator>
+#include <chrono>
+
+namespace wave {
+
+using TimeType = std::chrono::steady_clock::time_point;
+
+/** Measurement type to store in the container */
+template<typename T, typename S>
+struct Measurement {
+    TimeType time_point;
+    S sensor_id;
+    T value;
+
+    Measurement(const TimeType &t, const S &s, const T &v) :
+            time_point{t},
+            sensor_id{s},
+            value{v} {}
+};
+
+/** Perform linear interpolation between two measurements
+ *
+ * This template can be specialized for measurement types requiring a
+ * different calculation
+ */
+template<typename T>
+decltype(T::value) interpolate(const T &m1, const T &m2, const TimeType &t) {
+    auto w2 = 1.0 * (t - m1.time_point) / (m2.time_point - m1.time_point);
+    return (1 - w2) * m1.value + w2 * m2.value;
+};
+
+namespace internal {
+
+using boost::multi_index::member;
+using boost::multi_index::indexed_by;
+using boost::multi_index::ordered_unique;
+using boost::multi_index::ordered_non_unique;
+using boost::multi_index::composite_key;
+using boost::multi_index::tag;
+
+// This template sets up the internal boost_multi_index type for a given
+// measurement type T.
+template<typename T>
+struct measurement_container {
+    // Tags for accessing indices in the multi_index_container
+    struct time_index {};
+    struct sensor_index {};
+    struct composite_index {};
+
+    using SensorIdType = decltype(T::sensor_id);
+
+    // Define a container able to index by time, sensor id, or both
+    // First, define which members of the Measurement object are used as keys
+    struct time_key : member<T, TimeType, &T::time_point> {};
+    struct sensor_key : member<T, SensorIdType, &T::sensor_id> {};
+    // Define a composite key combining the above two keys
+    // Note the composite index is sorted by sensor id first, then time
+    struct sensor_and_time_key : composite_key<T, sensor_key, time_key> {};
+
+    // Define an index for each key. Each index will be accessible via its tag
+    struct indices : indexed_by<
+            ordered_non_unique<tag<time_index>, time_key>,
+            ordered_non_unique<tag<sensor_index>, sensor_key>,
+            ordered_unique<tag<composite_index>, sensor_and_time_key>> {
+    };
+    // Note we use separate struct definitions above, instead of typedefs, to
+    // reduce the length of the name printed by the compiler.
+
+    // Finally, define the multi_index_container type.
+    using type = boost::multi_index_container<T, indices, std::allocator<T>>;
+
+    // Get the type of the composite index, using its tag
+    using composite_type = typename type::template index<composite_index>::type;
```

Comment:
What on earth is this doing? I was just browsing out of interest, but this code is way beyond my understanding. I feel like this will be an issue going forward with code maintenance.